File: March 18 '22/Coin Piles gfg.cpp
class Solution {
  public:
    int minSteps(int a[], int n, int k) {
        sort(a, a+n);
        vector<int> pa(n+1, 0);
        
        for(int i=1; i<=n; i++)
            pa[i] = pa[i-1]+a[i-1];
            
        int res = 1e9;
        for(int i=0; i<n; i++) {
            int j = upper_bound(a,a+n,a[i]+k)-a;
            res = min(res, pa[i]+(pa[n]-pa[j]-(n-j)*(a[i]+k)));
        }
        return res;
    }
};

File: JUNE 2025/17 June/Coin Piles gfg.java
class Solution {
    public int minimumCoins(int[] arr, int k) {
        // code here
        Arrays.sort(arr);
        int n = arr.length;
        int ans = Integer.MAX_VALUE;
        int totalLeft = 0;
        
        for (int i = 0; i < n; ++i) {
            totalLeft += (i > 0) ? arr[i-1] : 0;
            int removeRight = 0;
            int limit = arr[i] + k;

            int low = i, high = n - 1, pos = n;
            while (low <= high) {
                int mid = (low + high) / 2;
                if (arr[mid] > limit) {
                    pos = mid;
                    high = mid - 1;
                } else {
                    low = mid + 1;
                }
            }

            for (int j = pos; j < n; ++j) {
                removeRight += arr[j] - limit;
            }

            ans = Math.min(ans, totalLeft + removeRight);
        }

        return ans;
    }
}
